Core: Add `reserve` function to `Array`, `Vector`, and `String`

Ivorforce opened the pull request:
- Closes https://github.com/godotengine/godot-proposals/issues/2954
- Requires https://github.com/godotengine/godot/pull/100145
- Follow-up of https://github.com/godotengine/godot/pull/100619
- (Mostly) Supersedes https://github.com/godotengine/godot/pull/74582

This PR adds `reserve()` to `CowData`, `Vector`, and `Array`. 
It achieves this by storing the capacity into the buffer, rather than inferring it from size.

This PR also refactors `CowData` almost completely. This was necessary for `reserve` to work properly, but has the side effect of neatly organizing functionality into smaller, specialized packages. This accelerates most functions.

## Explanation of behavior

`CowData` (and classes using it) currently have somewhat unusual behavior. This PR recreates current behavior, rather than changing it.
In particular:
- All element removal operations (e.g. `clear` and `remove_at` also shrink buffer. `LocalVector` for example only shrinks the buffer on explicit request (`reset`).
- Allocation sizes are deterministic by size (rather than depending on the number of objects being added or removed at once).
- Allocation size is grown by powers of 2, but measured as bytes rather than number of objects.

### New behavior

`CowData` is interesting in that there are usually 3 cases to account for (for size-changing functions). Each implies a different ideal strategy:
- There is no data yet.
    - -> Allocate new data, then add the elements.
- There is data, and we're the sole owner.
    - -> Do whatever is fastest on existing data.
- There is data, but there are other owners.
    - -> Allocate new data, copy elements directly into the spots where they're needed, and add new elements in the gaps.

If you review relevant functions, you should see most of them are structured in this way (with 3 separate branches). `resize` is an exception because it needs to handle both growth and shrinking efficiently.

## Explanation of code changes

I _really_, really wanted https://github.com/godotengine/godot/pull/100619 to be the large portion of work, and for this PR to be a small addition on top. 
Unfortunately, it was not possible, because current behavior of `CowData` necessitates differentiating harshly between _growth_ operations and _shrinking_ / _changing_ operations.
In particular, _growth_ operations (currently only `push_back`) are allowed to keep the previously reserved capacity, while _shrinking_ / _changing_ operations also shrink the backing buffer. It must be done this way to avoid regressions.

On the plus side, differentiating between growth and shrinking operations allows us to instantiate _far less_ code per call.
This shows in the final binary by the intense reduction of size.

I recommend reviewing this PR by looking at the result, rather than the changes. Almost all implementations needed to be changed to accomodate `reserve`, so trying to map them to previous behavior might be a bit of a waste of time.

Functions are documented by comments. If they're not enough, do not hesitate to ask! Names and comments are definitely open to discussion :)

## Benchmarks

I ran a pretty extensive benchmark to make sure there are no speed regressions. Some cases are marginally slower than before, but nothing is especially notable. In turn, a bunch of functions are noticeably faster than before.

The test code runs through several configurations of array size vs. iterations, and logs the time needed for each.

<details>

<summary>Test Code</summary>

```c++
	for (int64_t arr_size : { 4, 32, 512, 1024, 20000, 200000 }) {
		int64_t N = 200000000 / arr_size;
		int64_t N1 = 200000 / arr_size;
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N; run++) {
				auto t0 = std::chrono::high_resolution_clock::now();
				Vector<int64_t> vector;
				CRASH_COND(vector.resize(arr_size));
				int64_t *ptrw = vector.ptrw();
				for (int i = 0; i < arr_size; i ++) {
					// Test
					ptrw[i] = 5;
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "resize-ptrw: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N; run++) {
				auto t0 = std::chrono::high_resolution_clock::now();
				Vector<int64_t> vector;
				CRASH_COND(vector.resize(arr_size));
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.write[i] = 5;
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "resize-write: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N; run++) {
				auto t0 = std::chrono::high_resolution_clock::now();
				Vector<int64_t> vector;
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.insert(vector.size(), 5);
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "insert back: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N; run++) {
				auto t0 = std::chrono::high_resolution_clock::now();
				Vector<int64_t> vector;
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.push_back(5);
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "push_back: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N; run++) {
				auto t0 = std::chrono::high_resolution_clock::now();
				Vector<int64_t> vector;
				CRASH_COND(vector.reserve(arr_size));
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.push_back(5);
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "reserve push_back: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N1; run++) {
				auto t0 = std::chrono::high_resolution_clock::now();
				Vector<int64_t> vector;
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.insert(0, 5);
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "insert front: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N; run++) {
				Vector<int64_t> vector;
				CRASH_COND(vector.resize(arr_size));
				auto t0 = std::chrono::high_resolution_clock::now();
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.remove_at(vector.size() - 1);
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "remove_at back: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
		{
			size_t time_spent_ns = 0;

			for (int64_t run = 0; run < N1; run++) {
				Vector<int64_t> vector;
				CRASH_COND(vector.resize(arr_size));
				auto t0 = std::chrono::high_resolution_clock::now();
				for (int i = 0; i < arr_size; i ++) {
					// Test
					vector.remove_at(0);
				}
				auto t1 = std::chrono::high_resolution_clock::now();
				time_spent_ns += std::chrono::duration_cast<std::chrono::nanoseconds>(t1 - t0).count();

				vector.clear();
			}

			std::cout << "remove_at front: " << arr_size << std::endl;
			std::cout << time_spent_ns / 1000 / 1000 << "ms\n";
		}
	}
```
</details>

This yielded:

<details>

<summary>Results on this PR</summary>

```
resize-ptrw: 4
1442ms
resize-write: 4
1470ms
insert back: 4
4403ms
push_back: 4
4075ms
reserve push_back: 4
1582ms
insert front: 4
4ms
remove_at back: 4
3442ms
remove_at front: 4
3ms
resize-ptrw: 32
241ms
resize-write: 32
388ms
insert back: 32
1966ms
push_back: 32
1466ms
reserve push_back: 32
573ms
insert front: 32
2ms
remove_at back: 32
2297ms
remove_at front: 32
2ms
resize-ptrw: 512
28ms
resize-write: 512
165ms
insert back: 512
983ms
push_back: 512
601ms
reserve push_back: 512
423ms
insert front: 512
8ms
remove_at back: 512
1045ms
remove_at front: 512
8ms
resize-ptrw: 1024
22ms
resize-write: 1024
155ms
insert back: 1024
887ms
push_back: 1024
532ms
reserve push_back: 1024
418ms
insert front: 1024
16ms
remove_at back: 1024
947ms
remove_at front: 1024
14ms
resize-ptrw: 20000
32ms
resize-write: 20000
156ms
insert back: 20000
792ms
push_back: 20000
435ms
reserve push_back: 20000
423ms
insert front: 20000
250ms
remove_at back: 20000
841ms
remove_at front: 20000
249ms
resize-ptrw: 200000
139ms
resize-write: 200000
253ms
insert back: 200000
891ms
push_back: 200000
525ms
reserve push_back: 200000
510ms
insert front: 200000
3100ms
remove_at back: 200000
835ms
remove_at front: 200000
3085ms
```
</details>

<details>

<summary>Results on `master`</summary>

```
resize-ptrw: 4
1406ms
resize-write: 4
1521ms
push_back: 4
3764ms
insert front: 4
3ms
remove_at back: 4
3355ms
remove_at front: 4
3ms
resize-ptrw: 32
238ms
resize-write: 32
577ms
push_back: 32
2176ms
insert front: 32
2ms
remove_at back: 32
2564ms
remove_at front: 32
2ms
resize-ptrw: 512
27ms
resize-write: 512
402ms
push_back: 512
1331ms
insert front: 512
8ms
remove_at back: 512
1359ms
remove_at front: 512
6ms
resize-ptrw: 1024
22ms
resize-write: 1024
397ms
push_back: 1024
1240ms
insert front: 1024
15ms
remove_at back: 1024
1256ms
remove_at front: 1024
13ms
resize-ptrw: 20000
26ms
resize-write: 20000
395ms
push_back: 20000
1123ms
insert front: 20000
297ms
remove_at back: 20000
1158ms
remove_at front: 20000
242ms
resize-ptrw: 200000
133ms
resize-write: 200000
463ms
push_back: 200000
1191ms
insert front: 200000
3375ms
remove_at back: 200000
1155ms
remove_at front: 200000
3093ms
```
</details>

To summarize:
- `resize` then `ptrw`-assign is ~3% slower than before.
- `resize` then `.write`-assign is 2% to 50% faster than before.
- `push_back` is 10% slower than before (small arrays) to 2x faster than before (medium to large arrays)
- `reserve` then `push_back` is 3% (large arrays) to 3x (medium size arrays) faster than regular `push_back`.
- `push_back` is 10% to 2x faster than `insert` at back (which was the previous implementation)
- `insert` at front can be 10% faster than before.
- `remove_at` at back is 1.5% slower (small arrays) to 30% faster (large arrays) than before.

I also tested `reserve` from GDScript to make sure it's still interesting there. It seems it can make a small but respectable difference of 20%:

```gdscript
func _ready() -> void:
	test(5, false)
	test(5, true)
	test(50, false)
	test(50, true)
	test(500, false)
	test(500, true)
	test(5000, false)
	test(5000, true)

func test(n: int, reserve: bool):
	var start := Time.get_ticks_usec()

	@warning_ignore("integer_division")
	for r in 10000000 / n:
		var a := []
		if reserve:
			a.reserve(n)
		for i in n:
			a.push_back(0)

	var end := Time.get_ticks_usec()
	@warning_ignore("integer_division")
	print("%dms" % ((end - start) / 1000))
```

Printing:
```
687ms
590ms
463ms
386ms
382ms
364ms
371ms
364ms
```

_Edit: It was decided later to not expose `reserve` to GDScript_

RedMser commented:
Some thoughts regarding the documentation / scripting API:

- `reserve` and `reverse` visually look very similar and could easily be swapped. Maybe a more explicit name like `reserve_capacity` or similar could be found?
- Docs don't really clarify the difference between `reserve` and `resize`. To me, they seem really similar, so I wouldn't know when to use which one... They should maybe also link to each other, since they seem related.

Mickeon commented:
Doing god's work.

> * `reserve` and `reverse` visually look very similar and could easily be swapped. Maybe a more explicit name like `reserve_capacity` or similar could be found?

I do agree, but to play devil's advocate, one has a required parameter and the other does not even have any.

Mickeon commented:
> GDScript has a lot of overhead, so `reserve` has a lot less of a measurable effect there as opposed to core. Or at least that's what I suspect. I've been meaning to test this.

Even then, the expectation for Array in GDScript is that... it just works, more so than doing so efficiently. On the other hand, Packed arrays are generally expected to be fast (by extension of being compact) so the inclusion of this method sounds more justified there. Still, worth testing.

Ivorforce committed:
Store current capacity in `CowData` buffers, and rewrite most of it.
Add `reserve` to `CowData`, `Vector` and `Array`.

# Conflicts:
#	core/os/memory.h
#	core/templates/cowdata.h

Ivorforce commented:
I've run an extensive suite of performance tests to make sure there are only marginal regressions at worst, but improvements otherwise.

Most notably, `push_back` can be 2x faster than before as-is, and using `reserve` alongside it can be up to 6x faster (in total).

I also tested that it makes sense from GDScript, where I can see an up to 20% measurable improvement (in the best case). As expected, it's not _that_ interesting (as opposed to in `Core`), but it's at least _somewhat_ interesting.

DeeJayLSP commented on doc/classes/Array.xml:
```diff
@@ -669,6 +669,14 @@
+			<return type="int" />
+			<param index="0" name="min_capacity" type="int" />
+			<description>
+				Reserve space for at least [param min_capacity] total elements.
```
Seconding what was said in the discussion, I think it's important to mention what exactly this does and the benefits, like "if a certain amount of capacity is reserved, subsequent `push_backs` will be sped up until the size gets bigger than that".

Ivorforce replied:
Thanks for the feedback :)
We've reached a conclusion in the last core meeting to _not_ expose this function to GDScript for now, as it's not that useful in comparison (and potentially confusing against `resize`). So my next push will remove this change anyway!

DeeJayLSP commented on core/templates/cowdata.h:
```diff
@@ -209,30 +220,12 @@ class CowData {
-
-		resize(len - 1);
-	}
+	[[nodiscard]] Error reserve(USize p_min_capacity);
```
Is `[[nodiscard]` really needed? For most use cases, I think this just enforces unnecessary error handling.

Ivorforce replied:
A failed `reserve` allocation is almost certainly going to lead to undefined behavior afterwards (e.g. from failed `push_back` calls), so it would be good if callers need to consciously handle this.

Other, similar APIs such as `resize` currently don't force callers to handle this case, which I think is a mistake.  Of course, this is somewhat of a departure from our current API design, and it may turn out to be annoying for other developers. I still think it's the right call, but I'm open to discussing or changing this!

lawnjelly replied:
I also agree with @DeeJayLSP and wouldn't add the `[[nodiscard]]`.

Effectively if this happens, if this doesn't crash, something else will .. I think as an engine Godot has positively taken the decision to not handle out of memory conditions (in most cases), and so forcing call sites to add error handling code goes against this.

Ivorforce replied:
Ok, I made it non-nodiscard.

Calinou reviewed (approved):
Tested locally, it works as expected. I've tested lightmap baking in a large scene and I don't get a crash due to CowData size or something similar.

Code looks good to me.

## Benchmark

<details open>

<summary>PC specifications</summary>

- **CPU:** AMD Ryzen 9 9950X3D
- **GPU:** NVIDIA GeForce RTX 5090
- **RAM:** 64 GB (2×32 GB DDR5-6000 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Linux (Fedora 42)
</details>

Using the C++ code which I added at the top of `Main::setup()` (remember to `#include <iostream>` above the function).

<details>

```
resize-ptrw: 4
934ms
resize-write: 4
989ms
insert back: 4
2065ms
push_back: 4
1846ms
reserve push_back: 4
1047ms
insert front: 4
2ms
remove_at back: 4
1835ms
remove_at front: 4
1ms
resize-ptrw: 32
120ms
resize-write: 32
281ms
insert back: 32
1073ms
push_back: 32
701ms
reserve push_back: 32
316ms
insert front: 32
1ms
remove_at back: 32
896ms
remove_at front: 32
0ms
resize-ptrw: 512
17ms
resize-write: 512
226ms
insert back: 512
699ms
push_back: 512
317ms
reserve push_back: 512
229ms
insert front: 512
2ms
remove_at back: 512
684ms
remove_at front: 512
2ms
resize-ptrw: 1024
16ms
resize-write: 1024
194ms
insert back: 1024
631ms
push_back: 1024
284ms
reserve push_back: 1024
223ms
insert front: 1024
3ms
remove_at back: 1024
653ms
remove_at front: 1024
3ms
resize-ptrw: 20000
9ms
resize-write: 20000
216ms
insert back: 20000
626ms
push_back: 20000
242ms
reserve push_back: 20000
219ms
insert front: 20000
83ms
remove_at back: 20000
618ms
remove_at front: 20000
75ms
resize-ptrw: 200000
10ms
resize-write: 200000
215ms
insert back: 200000
615ms
push_back: 200000
230ms
reserve push_back: 200000
220ms
insert front: 200000
921ms
remove_at back: 200000
549ms
remove_at front: 200000
912ms
```
</details>

Ivorforce commented:
> Interesting warning on the Mingw-GCC template build:
> 
> ```
>  ERROR: In file included from ./core/templates/local_vector.h:34,
>                  from ./core/object/message_queue.h:35,
>                  from ./core/object/object.h:35,
>                  from ./core/variant/binder_common.h:34,
>                  from ./core/object/method_bind.h:33,
>                  from ./core/object/class_db.h:33,
>                  from ./core/object/ref_counted.h:33,
>                  from ./core/io/xml_parser.h:33,
>                  from ./core/doc_data.h:33,
>                  from ./core/object/script_language.h:33,
>                  from scene\gui\code_edit.h:33,
>                  from scene\gui\code_edit.cpp:31:
> In function 'void copy_arr_placement(T*, const T*, size_t) [with T = Pair<int, int>]',
>     inlined from 'CowData<T>::CowData(std::initializer_list<_Tp>) [with T = Pair<int, int>]' at ./core/templates/cowdata.h:574:20,
>     inlined from 'Vector<T>::Vector(std::initializer_list<_Tp>) [with T = Pair<int, int>]' at ./core/templates/vector.h:332:4,
>     inlined from 'ScriptLanguage::CodeCompletionOption::CodeCompletionOption()' at ./core/object/script_language.h:322:26,
>     inlined from 'List<T, A>::Element::Element() [with T = ScriptLanguage::CodeCompletionOption; A = DefaultAllocator]' at ./core/templates/list.h:138:28,
>     inlined from 'List<T, A>::Element* List<T, A>::push_back(const T&) [with T = ScriptLanguage::CodeCompletionOption; A = DefaultAllocator]' at ./core/templates/list.h:297:16,
>     inlined from 'void CodeEdit::add_code_completion_option(CodeCompletionKind, const String&, const String&, const Color&, const Ref<Resource>&, const Variant&, int)' at scene\gui\code_edit.cpp:2238:44:
> ./core/os/memory.h:221:23: error: using a dangling pointer to 'C.1282' [-Werror=dangling-pointer=]
>   221 |                 memcpy((uint8_t *)p_dst, (uint8_t *)p_src, p_num * sizeof(T));
>       |                 ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./core/object/script_language.h: In member function 'void CodeEdit::add_code_completion_option(CodeCompletionKind, const String&, const String&, const Color&, const Ref<Resource>&, const Variant&, int)':
> ./core/object/script_language.h:322:40: note: 'C.1282' declared here
>   322 |                 CodeCompletionOption() {}
>       |                                        ^
> ```

I'm 90% sure this is a [false positive](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=109514). GCC really doesn't seem to understand this code. I've added an ignore with the others at the top of the file.

lawnjelly commented on core/templates/cowdata.h:
```diff
@@ -272,139 +265,289 @@ void CowData<T>::_unref() {
+		*_get_size() = new_size;
+	} else {
+		// Remove by forking.
+		CRASH_COND(_copy_to_new_buffer(new_size, p_index, 0, new_size - p_index));
```
I would change to this, both because clearer, and also because it's possible someone will disable `CRASH_COND()` macro, in which case, it won't run at all.
```
Error err = _copy_to_new_buffer(new_size, p_index, 0, new_size - p_index);
CRASH_COND(err != OK);
```

lawnjelly commented on core/templates/cowdata.h:
```diff
@@ -272,139 +265,289 @@ void CowData<T>::_unref() {
+			const USize new_alloc_size = _get_alloc_size(p_size);
+			const USize prev_alloc_size = _get_alloc_size(capacity());
+			if (new_alloc_size < prev_alloc_size) {
+				_ALLOW_DISCARD_ _realloc_bytes(new_alloc_size); // Always succeeds on shrink.
```
Not that I'm suggesting it is wise to do much alloc failure handling, but is there any guarantee that a realloc to a smaller size can't fail?
Grok suggests it is less likely than a grow, but it is not guaranteed.

If it can shrink by e.g. decommiting pages or another method that is one thing, but if it has to create a smaller buffer then copy, it's still allocating more than was there originally?

lawnjelly reviewed (approved):
Apart from a few comments this looks great. I've only looked through the `CowData` and `Vector` in detail, but those are the important parts (I ran out of time).

@Ivorforce today alerted me to the really awful behaviour of current `Vector` when resizing on the border of allocation size differences, it shrinks and grows every time you add / remove a single element.

Adding a capacity is therefore essential so we can do hysteresis and avoid these really bad situations (and possibly also the option of shrinking being manual only).

AThousandShips commented on core/os/memory.h:
```diff
@@ -36,6 +36,11 @@
 #include <new> // IWYU pragma: keep // `new` operators.
 #include <type_traits>
 
+static constexpr size_t mem_aligned_address(size_t address, size_t alignment) {
```
```suggestion
static constexpr size_t mem_aligned_address(size_t p_address, size_t p_alignment) {
```

AThousandShips commented on core/variant/array.h:
```diff
@@ -135,6 +135,7 @@ class Array {
 	_FORCE_INLINE_ void append(const Variant &p_value) { push_back(p_value); } //for python compatibility
 	void append_array(const Array &p_array);
 	Error resize(int p_new_size);
+	Error reserve(int64_t p_new_size);
```
My only concern here is that the reserve function supports a wider range than the resize, for uniformity I'd suggest using `int` to not create strange cases, you could easily have cases where you do `reserve(size)` followed by operations adding elements or resizing triggering errors or weirdness as the size isn't supported

Repiteo reviewed (approved):
Approved in core meeting

Repiteo commented:
Needs rebase after #100145

Repiteo commented:
Thanks!
